storage: bump Pebble and add cluster settings for compaction concurrency

RaduBerinde opened the pull request:
#### go.mod: bump Pebble to 16f3b5764d35

Changes:

 * [`16f3b576`](https://github.com/cockroachdb/pebble/commit/16f3b576) db: change MaxConcurrentCompactions() to return a range

Release note: none.
Epic: none.

#### storage: fix up compaction concurrency code


#### storage: add cluster settings for compaction concurrency

Add two cluster settings that control the compaction concurrency
limits: `storage.concurrent_compactions` and
`storage.max_concurrent_compactions`.

Fixes: #144963
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/145301)
<!-- Reviewable:end -->

RaduBerinde committed:
storage: add cluster settings for compaction concurrency

Add two cluster settings that control the compaction concurrency
limits: `storage.compaction_concurrency` and
`storage.max_compaction_concurrency`.

Note that the `COCKROACH_CONCURRENT_COMPACTIONS` env var is still
supported: when it is used together with the cluster setting, we take
the max of the two values.

Fixes: #144963
Release note: None

RaduBerinde committed:
storage: fix up compaction concurrency code

RaduBerinde committed:
go.mod: bump Pebble to e90517277fca

Changes:

 * [`e9051727`](https://github.com/cockroachdb/pebble/commit/e9051727) db: change MaxConcurrentCompactions() to return a range
 * [`7c06a360`](https://github.com/cockroachdb/pebble/commit/7c06a360) db: add option for CompactionGarbageFractionForMaxConcurrency

Release note: none.
Epic: none.

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 3 of 4 files at r1, 5 of 5 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/145301#-:-OP3RQYMC4BPVREA0-rE:bg7igkv)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @jbowens and @xxmplus)

___
*[`pkg/storage/pebble.go` line 676 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/145301#-OP3QvgT4w87DFrNdk-y:-OP3QvgT4w87DFrNdk-z:bhwxecg) ([raw file](https://github.com/cockroachdb/cockroach/blob/a13d99614ac9483fa717bfe0a209d2155ef87ac1/pkg/storage/pebble.go#L676)):*
> ```Go
> 	}
> 	// Note: the MaxConcurrentCompactions function will be wrapped below to allow
> 	// overriding dynamically via overriding the max at runtime through
> ```

nit: ... the min and max ...
___
*[`pkg/storage/pebble.go` line 794 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/145301#-OP3QsuV-W9hHpyTuLun:-OP3QsuV-W9hHpyTuLuo:bhwxecg) ([raw file](https://github.com/cockroachdb/cockroach/blob/a13d99614ac9483fa717bfe0a209d2155ef87ac1/pkg/storage/pebble.go#L794)):*
> ```Go
> 
> 	// Wrap the MaxConcurrentCompactions function to allow overriding dynamically via
> 	// overriding the max at runtime through Engine.SetCompactionConcurrency.
> ```

nit: ... the min and max ...


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/145301#-:-OP3clZ-2OUzBBA0nq3D:b-814g8k)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @jbowens and @xxmplus)



<!-- Sent from Reviewable.io -->

jbowens reviewed (approved):
> Reviewed 4 of 4 files at r1, 5 of 5 files at r2, 1 of 1 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/145301)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @xxmplus)

___
*[`pkg/storage/pebble.go` line 296 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/145301#-OP6BKuq0jvgXyVsHvjD:-OP6BKuq0jvgXyVsHvjE:b-nd795l) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ed47a7b67bbae4519c7f5ebff199d4698f7b7d7/pkg/storage/pebble.go#L296)):*
> ```Go
> 	settings.ApplicationLevel,
> 	"storage.max_compaction_concurrency",
> 	"the maximum number of concurrent compactions (0 = default)",
> ```

I think we should document somewhere that this setting takes precedence over the environment variable `COCKROACH_CONCURRENT_COMPACTIONS`. There's quite a hierarchy now

1. In the absence of any configuration, we'll set max concurrent compactions to `min(GOMAXPROCS-1, 3)`.
2. If `COCKROACH_ROCKSDB_CONCURRENCY` is set, it's next in the hierarchy, and we set the max concurrent compactions to `COCKROACH_ROCKSDB_CONCURRENCY-1`.
3. If `COCKROACH_CONCURRENT_COMPACTIONS` is set, we prefer its value.
4. If `storage.max_compaction_concurrency` cluster setting is set, we prefer its value.
5. If a call to the crdb_internal func is calling `Engine.SetCompactionConcurrency`, we use its value.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
Interesting. The env var is not great in that we need to restart the node to change it. I'd hate to have to restart a node just to clear out its env var. We could take the max of the two settings - I'd guess in most cases we use the variable to increase the concurrency, not decrease it.

Then the default value for the cluster setting can actually be exactly 3, which makes things cleaner. (do we really need the min(GOMAXPROCS-1) thing? Do we ever expect crdb to run on nodes <= 4 CPUs? In that exceptional case we could just recommend changing the cluster setting to 1)

Note that`storage.compaction\_concurrency` (the lower bound) can be used to unconditionally increase the concurrency for the entire cluster (it also effectively overrides the max if the max is smaller).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/145301#-:-OP6OyWh24S0pW_K9zyh:b-67xjio)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @jbowens and @xxmplus)

___
*[`pkg/storage/pebble.go` line 296 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/145301#-OP6BKuq0jvgXyVsHvjD:-OP6OjbC7NnyoD7RW-tS:b8tq58e) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ed47a7b67bbae4519c7f5ebff199d4698f7b7d7/pkg/storage/pebble.go#L296)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

I think we should document somewhere that this setting takes precedence over the environment variable `COCKROACH_CONCURRENT_COMPACTIONS`. There's quite a hierarchy now

1. In the absence of any configuration, we'll set max concurrent compactions to `min(GOMAXPROCS-1, 3)`.
2. If `COCKROACH_ROCKSDB_CONCURRENCY` is set, it's next in the hierarchy, and we set the max concurrent compactions to `COCKROACH_ROCKSDB_CONCURRENCY-1`.
3. If `COCKROACH_CONCURRENT_COMPACTIONS` is set, we prefer its value.
4. If `storage.max_compaction_concurrency` cluster setting is set, we prefer its value.
5. If a call to the crdb_internal func is calling `Engine.SetCompactionConcurrency`, we use its value.
</blockquote></details>

Good call, added.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 1 of 4 files at r1, 3 of 3 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/145301#-:-OP6tTDH1h5f2EDMajvW:bg7igkv)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @jbowens and @xxmplus)

___
*[`pkg/storage/pebble.go` line 288 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/145301#-OP6sV3wEPKYDvBzvrIt:-OP6sV3xDP2-0OwTTMuG:bytyk45) ([raw file](https://github.com/cockroachdb/cockroach/blob/96082a03c6fbf392428c3a3ceacaf7a6f5a4a690/pkg/storage/pebble.go#L288)):*
> ```Go
> // precedence (i.e. the max concurrency will also equal this value).
> //
> // This value is temporarily overridden while
> ```

The "This value is temporarily overridden" sounds confusing to me -- it suggests something is temporarily overriding a cluster setting.
Perhaps make this comment reference the other comment that explains how these different ways of setting the concurrency interact.
___
*[`pkg/storage/pebble.go` line 311 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/145301#-OP6ssJ11BM5oJrF63sa:-OP6ssJ11BM5oJrF63sb:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/96082a03c6fbf392428c3a3ceacaf7a6f5a4a690/pkg/storage/pebble.go#L311)):*
> ```Go
> // the two.
> //
> // In all cases, this value is temporarily overridden while
> ```

ditto
___
*[`pkg/storage/pebble.go` line 839 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/145301#-OP6t7cwB8dujV3f_V1A:-OP6t7cwB8dujV3f_V1B:b-bodt3n) ([raw file](https://github.com/cockroachdb/cockroach/blob/96082a03c6fbf392428c3a3ceacaf7a6f5a4a690/pkg/storage/pebble.go#L839)):*
> ```Go
> 
> 	// Wrap the CompactionConcurrencyRange function to allow overriding the lower
> 	// and upper values at runtime through and the max at runtime through
> ```

nit: "and the max at runtime through" is repetitive


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/145301#-:-OP6veay6inyyTdh3swp:b-814g8k)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @jbowens, @sumeerbhola, and @xxmplus)

___
*[`pkg/storage/pebble.go` line 288 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/145301#-OP6sV3wEPKYDvBzvrIt:-OP6uuwS2iJ84uBo4Vnu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/96082a03c6fbf392428c3a3ceacaf7a6f5a4a690/pkg/storage/pebble.go#L288)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

The "This value is temporarily overridden" sounds confusing to me -- it suggests something is temporarily overriding a cluster setting.
Perhaps make this comment reference the other comment that explains how these different ways of setting the concurrency interact.
</blockquote></details>

Done.
___
*[`pkg/storage/pebble.go` line 311 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/145301#-OP6ssJ11BM5oJrF63sa:-OP6vXWR5Dwht_ko-ux5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/96082a03c6fbf392428c3a3ceacaf7a6f5a4a690/pkg/storage/pebble.go#L311)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

jbowens reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 3 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/145301#-:-OP7FlJ2CUV0ldXZMcz0:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @sumeerbhola and @xxmplus)



<!-- Sent from Reviewable.io -->

RaduBerinde commented:
TFTRs!

bors r+
